src/injected/util.ts

```diff
@@ -16,3 +18,31 @@ export const createFunction = (
   createdFunctions.set(functionValue, fn);
   return fn;
 };
+
+/**
+ * @internal
+ */
+export const checkVisibility = (
+  node: Node | null,
+  visible?: boolean
+): Node | boolean => {
+  if (!node) {
+    return visible === false;
+  }
+  if (visible === undefined) {
+    return node;
+  }
+  const element = (
+    node.nodeType === Node.TEXT_NODE ? node.parentElement : node
+  ) as Element;
+
+  const style = window.getComputedStyle(element);
+  const isVisible =
+    style && style.visibility !== 'hidden' && isBoundingBoxVisible(element);
+  return visible === isVisible ? node : false;
+};
+
+function isBoundingBoxVisible(element: Element): boolean {
+  const rect = element.getBoundingClientRect();
+  return !!(rect.top || rect.bottom || rect.width || rect.height);
```

Comment:
I guess there is a bug here? {top: 1, bottom: 1, width: 0, height: 0} would be treated as visible? Let's add unit tests for these functions why we are at it?

Reply:
Fixed in https://github.com/puppeteer/puppeteer/pull/8954